File: util/common/src/main/java/com/common/frame/zkpath/ZkActiveValue.java
package com.common.frame.zkpath;

import com.common.util.SysUtil;

public class ZkActiveValue {
	private String hostOrIp=null;//主机名或主机ip
	private String procId=null;//系统进程标识，系统产生
	private String channelId="";//通道id，即进程实例标识
	public ZkActiveValue() {
		
	}
	public ZkActiveValue(String channelId) {
		hostOrIp= SysUtil.getHostName();
		procId=SysUtil.getProcessId();
		this.channelId=channelId;
	}
	public String getHostOrIp() {
		return hostOrIp;
	}

	public void setHostOrIp(String hostOrIp) {
		this.hostOrIp = hostOrIp;
	}

	public String getChannelId() {
		return channelId;
	}
	public void setChannelId(String channelId) {
		this.channelId = channelId;
	}
	public String getProcId() {
		return procId;
	}

	public void setProcId(String procId) {
		this.procId = procId;
	}
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((channelId == null) ? 0 : channelId.hashCode());
		result = prime * result + ((hostOrIp == null) ? 0 : hostOrIp.hashCode());
		result = prime * result + ((procId == null) ? 0 : procId.hashCode());
		return result;
	}
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (!(obj instanceof ZkActiveValue))
			return false;
		ZkActiveValue other = (ZkActiveValue) obj;
		if (channelId == null) {
			if (other.channelId != null)
				return false;
		} else if (!channelId.equals(other.channelId))
			return false;
		if (hostOrIp == null) {
			if (other.hostOrIp != null)
				return false;
		} else if (!hostOrIp.equals(other.hostOrIp))
			return false;
		if (procId == null) {
			if (other.procId != null)
				return false;
		} else if (!procId.equals(other.procId))
			return false;
		return true;
	}
	@Override
	public String toString() {
		return "{\"hostOrIp\":\"" + hostOrIp + "\",\"procId\":\"" + procId + "\",\"channelId\":\"" + channelId + "\"} ";
	}
		
}

File: zkClient(C++)/frame/ZkActiveValue.cpp
#include <unistd.h>
#include <stdio.h>
#include <sstream>

#include "ZkEnum.h"
#include "ZkActiveValue.h"

ZkActiveValue::ZkActiveValue()
{
	hostOrIp ="";
	procId   ="";
	channelId="";
}

ZkActiveValue::ZkActiveValue(const ZkActiveValue& right)
{
  *this = right;
}

ZkActiveValue::ZkActiveValue(const string channelId)
{
	char hostname[32];
	if( gethostname(hostname,sizeof(hostname)) )
	{
	  this->hostOrIp  = "localhost";
	  #ifdef _DEBUG_
      cout<<"gethostname fail!"<<__FILE__<<__LINE__<<endl;
    #endif
	}else{
		this->hostOrIp  = hostname;
	}

	stringstream stream;
	stream<<getpid();

	this->procId    = stream.str();
	this->channelId = channelId;
	
	stream.clear();
	stream.str("");
}

ZkActiveValue::~ZkActiveValue()
{
}

void ZkActiveValue::operator = (const ZkActiveValue& right)
{
  if(this != &right)
  {
  	hostOrIp  = right.hostOrIp;
		procId    = right.procId;
		channelId = right.channelId;
  }
}

bool operator==(const ZkActiveValue &left,const ZkActiveValue& right)
{
  return (left.hostOrIp == right.hostOrIp && 
  			  left.procId 	== right.procId 	&&
  			  left.channelId== right.channelId);
}

bool operator<(const ZkActiveValue &left,const ZkActiveValue& right)
{
	if(left.hostOrIp < right.hostOrIp) return true;
  if(left.hostOrIp > right.hostOrIp) return false;
  	
  if(left.procId < right.procId) return true;
  if(left.procId > right.procId) return false;			

  return (left.channelId < right.channelId);
}

std::ostream& operator<<(std::ostream& os,const ZkActiveValue& po)
{
  os<<"{hostOrIp:"<< po.hostOrIp	<<","<<
      "procId:"		<< po.procId		<<","<<
      "channelId:"<< po.channelId	<<"}"<<endl;
  return os;
}


string ZkActiveValue::getHostOrIp() {
	return hostOrIp;
}

void ZkActiveValue::setHostOrIp(string hostOrIp) {
	this->hostOrIp = hostOrIp;
}

string ZkActiveValue::getChannelId() {
	return channelId;
}

void ZkActiveValue::setChannelId(string channelId) {
	this->channelId = channelId;
}

string ZkActiveValue::getProcId() {
	return procId;
}

void ZkActiveValue::setProcId(string procId) {
	this->procId = procId;
}